### PR TITLE
[Kernel] Define LogReplay assertLogFilesBelongToTable() helper method

### PR DESCRIPTION
**Pull request by PatrickLeahey**
### Which Delta project/connector is this regarding?
- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [x] Kernel
- [ ] Other (fill in here)

### Description
Resolves https://github.com/delta-io/delta/issues/2150

Defines LogReplay assertLogFilesBelongToTable() helper method based on the Spark implementation [here](https://github.com/delta-io/delta/blob/4f9c8b9cc294ec7b321847115bf87909c356bc5a/spark/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala#L430) as specified in issue.

Verifies that a set of delta or checkpoint files to be read actually belong to the transaction log of the delta table. Throws AssertionError if violation detected.

### CODE REVIEW

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java
Comment (vkorukanti) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:248]: We are planning to remove the `Path` as the code is copied from the Hadoop and we don't want to maintain a copy.

Wondering can you directly use the prefix search? Both the `logPath` and `FileStatus.getPath` are fully resolved.

File: kernel/kernel-api/src/test/java/io/delta/kernel/internal/replay/TestLogReplay.java
Comment (vkorukanti) [inline kernel/kernel-api/src/test/java/io/delta/kernel/internal/replay/TestLogReplay.java:26]: For tests we made a decision to write them in Scala, because the tests can be very concise in Scala. Could you please convert this into a Scala suite. You can find examples in the `kernel-defaults` module or `kernel-api` module.

**Review by vkorukanti — COMMENTED** [review]
Looks great, thank you for the contribution.

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java
Comment (PatrickLeahey) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:248]: Thanks for the feedback. To clarify, are you saying that the plan is to remove the `Path` class entirely? What will the logPath parameter type change to?

File: kernel/kernel-api/src/test/java/io/delta/kernel/internal/replay/TestLogReplay.java
Comment (PatrickLeahey) [reply kernel/kernel-api/src/test/java/io/delta/kernel/internal/replay/TestLogReplay.java]: Just rewrote tests in Scala

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java
Comment (vkorukanti) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:248]: Yep you are right, it is passed around a lot, but created in just few places. 

For now keep the `Path` as input arg and use the `path.toUri.toString` (create once in the beginning) for comparison.

File: kernel/kernel-api/src/test/scala/io/delta/kernel/internal/replay/LogReplaySuite.scala
Comment (vkorukanti) [inline kernel/kernel-api/src/test/scala/io/delta/kernel/internal/replay/LogReplaySuite.scala:30]: can you add a test where the paths are fully qualified e.g `s3://bucket/path/`

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java
Comment (PatrickLeahey) [reply kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java]: Should be resolved, we will see if tests pass

**Comment by PatrickLeahey** [comment]
@vkorukanti resolved your comments - thank you for the feedback!

File: kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java
Comment (sherlockbeard) [inline kernel/kernel-api/src/main/java/io/delta/kernel/internal/replay/LogReplay.java:251]: better to remove the ```Please contact Databricks Support.```  part .

**Review by vkorukanti — APPROVED** [review]

**Comment by PatrickLeahey** [comment]
@vkorukanti Feel free to dismiss other reviewers if you are ready to merge. Thank you for the review!

**Comment by vkorukanti** [comment]
@PatrickLeahey Will merge this PR once https://github.com/delta-io/delta/pull/2212 lands. It is adding lot of more tests and will verify whether anything break.

**Comment by PatrickLeahey** [comment]
@vkorukanti Looks like we are still good to go after #2212 merged

**Comment by PatrickLeahey** [comment]
> @PatrickLeahey Will merge this PR once #2212 lands. It is adding lot of more tests and will verify whether anything break.

✅

**Pull request merged**
